### ISSUE TITLE
Library use: Finding the range of lines that a method occupies in a file

### ISSUE THREAD

**Issue by oxinabox** [timholy/Revise.jl#236]
@timholy, I was rewatching your juliacon talk on Revise and Rebugger.
I recalled the part where you said 
"Revise is generalyl a useful too for anyone who wants to transform between representations of code."

And sure indeed it is, thank you.
eg: https://github.com/oxinabox/MagneticReadHead.jl/blob/ce566a86c8360d474f6448f034ce0543b890f107/src/breakpoints.jl#L13-L21

One thing I am yet to work out is how to find out what region in a file a method occupies.
I am pretty sure Revise knows this,
since Rebugger can pull the text content out (or maybe the function is in Rebugger?)
Obviously with just base julia I can find the line it starts from.
And will a bit of playing around I can thus break the any file up into regions.
But why duplicate that work, if you've already done it?

**Comment by timholy** [comment]
Yes. The key information is [here](https://github.com/timholy/Revise.jl/blob/97d3560783176c6ec32b9e660aee51a8e333c455/src/types.jl#L18). It looks more complicated than it is (i.e., all the `Union`s) because of "bootstrap" issues: if you add a new module or type to a package then the signatures may not be computable until you've `eval`ed the code.

A demo might make it more straightforward:

```julia
julia> using Revise, ColorTypes

julia> methods(red)
# 4 methods for generic function "red":
[1] red(c::ARGB32) in ColorTypes at /home/tim/.julia/packages/ColorTypes/BsAWO/src/traits.jl:17
[2] red(c::RGB24) in ColorTypes at /home/tim/.julia/packages/ColorTypes/BsAWO/src/traits.jl:16
[3] red(c::AbstractRGB) in ColorTypes at /home/tim/.julia/packages/ColorTypes/BsAWO/src/traits.jl:14
[4] red(c::TransparentColor{C,T,4} where T where C<:AbstractRGB) in ColorTypes at /home/tim/.julia/packages/ColorTypes/BsAWO/src/traits.jl:15

julia> pkgdata = Revise.pkgdatas[Base.PkgId(ColorTypes)]
Revise.PkgData(ColorTypes [3da002f7-5984-5a60-b8a6-cbb66c0b333f], "/home/tim/.julia/packages/ColorTypes/BsAWO", Dict("src/conversions.jl"=>FileInfo(OrderedCollections.OrderedDict{Module,Revise.FMMaps}(), "/home/tim/.julia/compiled/v1.1/ColorTypes/db21U.ji"),"src/ColorTypes.jl"=>FileInfo(OrderedCollections.OrderedDict{Module,Revise.FMMaps}(), "/home/tim/.julia/compiled/v1.1/ColorTypes/db21U.ji"),"src/types.jl"=>FileInfo(OrderedCollections.OrderedDict{Module,Revise.FMMaps}(), "/home/tim/.julia/compiled/v1.1/ColorTypes/db21U.ji"),"src/show.jl"=>FileInfo(OrderedCollections.OrderedDict{Module,Revise.FMMaps}(), "/home/tim/.julia/compiled/v1.1/ColorTypes/db21U.ji"),"src/traits.jl"=>FileInfo(OrderedCollections.OrderedDict{Module,Revise.FMMaps}(), "/home/tim/.julia/compiled/v1.1/ColorTypes/db21U.ji"),"src/operations.jl"=>FileInfo(OrderedCollections.OrderedDict{Module,Revise.FMMaps}(), "/home/tim/.julia/compiled/v1.1/ColorTypes/db21U.ji")))

julia> def = Revise.get_def(first(methods(red)))
:(red(c::ARGB32) = begin
          N0f8(c.color & 0x00ff0000 >> 16, 0)
      end)

julia> fmm = pkgdata.fileinfos["src/traits.jl"].fm[ColorTypes]
FMMaps(<158 expressions>, <143 signatures>)

julia> sigs, lineoffset = fmm.defmap[def]
(Any[Tuple{typeof(red),ARGB32}], 0)
```

The `lineoffset` encodes the current offset from when the method was compiled. In this case it's 0 because we haven't modified the `red` method. (In such cases you can just look up the line offset from the method itself.)

It's also worth pointing out that for parsing stacktraces, Rebugger has some extra code for extracting inlined methods: https://github.com/timholy/Rebugger.jl/blob/bee5dd42fc47950383fa8f474d3b62efd6bbf4a9/src/debug.jl#L56-L79, which is used https://github.com/timholy/Rebugger.jl/blob/bee5dd42fc47950383fa8f474d3b62efd6bbf4a9/src/debug.jl#L93-L113. I've sometimes wondered if that should be moved to Revise.

**Comment by timholy** [comment]
The query functionality is probably going to be split into a separate package, https://github.com/timholy/CodeTracking.jl. Consider opening PRs/issues for fleshing out exactly what kind of "query" support you'd like to see in CodeTracking. (A PR can be just a stub implementation of an API, since the package is nothing but stub implementations for now.)

Motivation discussed in https://github.com/JuliaDebug/ASTInterpreter2.jl/pull/37#issuecomment-457194413

**Comment by timholy** [comment]
Closed by #245 and #243
